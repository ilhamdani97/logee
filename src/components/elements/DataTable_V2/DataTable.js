import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scoped.css';
import Loader from '../../elements/Loader';
import chevronRight from '../../../assets/chevron-right-black.svg';
import chevronActive from '../../../assets/chevron-active.svg';

export default function DataTable(props) {
  const {
    className,
    column,
    data,
    emptyDataHandler,
    meta,
    isLoading,
    handleClick,
    children,
    identifier
  } = props;
  const classes = [styles.root, className].filter(Boolean).join(' ');
  const [openedChildren, setOpenedChildren] = useState('');
  const _showHideChild = item => {
    const id = item[identifier];
    setOpenedChildren(prevState => prevState === id ? '' : id);
    handleClick(item);
  };

  return (
    <>
      <table className={classes}>
        <thead>
          <tr>
            {children ? <th /> : <th>No</th>}
            {column.map((item, idx) => <TableHeader item={item} key={idx} />)}
          </tr>
        </thead>
        <tbody>
          {isLoading &&
            <tr className={styles.loading}><td colSpan={column.length + 1}><Loader /></td></tr>
          }
          {!isLoading && data && data.map((item, idx) => (
            <React.Fragment key={idx}>
              <TableRow
                activeRow={openedChildren}
                children={children}
                className={item.className || ''}
                column={column}
                handleClick={children ? () => _showHideChild(item) : handleClick}
                identifier={identifier}
                idx={idx}
                item={item}
                meta={meta}
              />
              {(openedChildren === item[identifier]) && (
                <>
                  <tr className={styles['child-row']} rowSpan={2}>
                    <td colSpan={column.length + 2}>
                      <div>
                        {children}
                      </div>
                    </td>
                  </tr>
                  <tr className={styles['empty-row']}><td /></tr>
                </>
              )}
            </React.Fragment>
          ))}
        </tbody>
      </table>
      {
        emptyDataHandler ?
          (data && !data[0]) &&
          emptyDataHandler :
          (data && !data[0]) &&
          <div className={styles.notFound}>
            <img src="../../assets/ic-not-found.svg" />
          </div>
      }
    </>
  );
}

DataTable.defaultProps = {
  children: null,
  className: '',
  column: [],
  data: [],
  emptyDataHandler: null,
  handleClick: () => { },
  identifier: 'id',
  isLoading: false,
  meta: {},
};

DataTable.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  column: PropTypes.array,
  data: PropTypes.array,
  emptyDataHandler: PropTypes.object,
  handleClick: PropTypes.func,
  identifier: PropTypes.string,
  isLoading: PropTypes.bool,
  meta: PropTypes.object
};

export function TableHeader({ item }) {

  return (
    <th className={styles['table-header']}>{item.heading}</th>
  );
}

TableHeader.defaultProps = {
  item: {},
};

TableHeader.propTypes = {
  item: PropTypes.object,
};

export function TableRow(props) {
  const {
    className,
    column,
    idx,
    item,
    meta,
    handleClick,
    children,
    activeRow,
    identifier
  } = props;
  const { page } = meta;
  const size = 10;
  let b = idx + 1;
  if (page > 1) {
    b = (idx + 1) + ((page - 1) * size);
  }
  return (
    <tr className={className} onClick={() => handleClick(item)}>
      {children ? (
        <td className={styles.arrow}>
          {activeRow === item[identifier] ? (
            <img src={chevronActive} />
          ) : (
            <img src={chevronRight} />
          )}
        </td>
      ) : <td>{b < 10 ? `0${b}` : b}</td>}
      {column.map((cItem, cIdx) => {
        const { value } = cItem;
        const newValue = typeof value === 'function' ? value(item, idx) : item[value];

        return (
          <td key={cIdx} title={newValue}>
            {newValue || '-'}
          </td>
        );
      })}
    </tr>
  );
}

TableRow.defaultProps = {
  activeRow: '',
  children: null,
  className: '',
  column: [],
  handleClick: () => { },
  identifier: '',
  idx: 0,
  item: {},
  meta: {}
};

TableRow.propTypes = {
  activeRow: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  column: PropTypes.array,
  handleClick: PropTypes.func,
  identifier: PropTypes.string,
  idx: PropTypes.number,
  item: PropTypes.object,
  meta: PropTypes.object,
};
