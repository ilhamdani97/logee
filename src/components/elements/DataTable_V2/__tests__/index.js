import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import DataTable, { TableHeader, TableRow } from '../DataTable';

const column = [
  { heading: 'Nama', value: 'name' },
  { heading: 'Email', value: 'email' },
  { heading: 'No. Handphone', value: 'phoneNumber' },
];

const data = [
  { name: 'Test', email: 'test@a.id', phoneNumber: '0811111111' },
  { name: 'Test', email: 'test@a.id', phoneNumber: '0811111111' },
];

describe('src/components/elements/DataTable', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<DataTable column={column} data={data} />);
    expect(tree).toMatchSnapshot();

    const result = DataTable({ ...DataTable.defaultProps, data: [], });
    expect(result.props.children[0].type).toBe('table');
    DataTable.defaultProps.handleClick();
  });

  test('render with children', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<DataTable children={<div/>} column={column} data={data}/>);
    expect(tree).toMatchSnapshot();
    const tbody = tree.props.children[0].props.children[1];
    const row = tbody.props.children[1][0].props.children[0];
    row.props.handleClick();
  });

  test('TableHeader', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<TableHeader />);
    expect(tree).toMatchSnapshot();
  });

  test('TableRow', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<TableRow column={column} />);
    expect(tree).toMatchSnapshot();
    TableRow.defaultProps.handleClick();
  });
});
