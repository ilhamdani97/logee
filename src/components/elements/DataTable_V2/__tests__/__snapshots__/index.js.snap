// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/DataTable TableHeader 1`] = `<th />`;

exports[`src/components/elements/DataTable TableRow 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td>
    01
  </td>
  <td>
    -
  </td>
  <td>
    -
  </td>
  <td>
    -
  </td>
</tr>
`;

exports[`src/components/elements/DataTable render 1`] = `
<React.Fragment>
  <table
    className=""
  >
    <thead>
      <tr>
        <th>
          No
        </th>
        <TableHeader
          item={
            Object {
              "heading": "Nama",
              "value": "name",
            }
          }
        />
        <TableHeader
          item={
            Object {
              "heading": "Email",
              "value": "email",
            }
          }
        />
        <TableHeader
          item={
            Object {
              "heading": "No. Handphone",
              "value": "phoneNumber",
            }
          }
        />
      </tr>
    </thead>
    <tbody>
      <React.Fragment>
        <TableRow
          activeRow=""
          className=""
          column={
            Array [
              Object {
                "heading": "Nama",
                "value": "name",
              },
              Object {
                "heading": "Email",
                "value": "email",
              },
              Object {
                "heading": "No. Handphone",
                "value": "phoneNumber",
              },
            ]
          }
          handleClick={[Function]}
          identifier="id"
          idx={0}
          item={
            Object {
              "email": "test@a.id",
              "name": "Test",
              "phoneNumber": "0811111111",
            }
          }
          meta={Object {}}
        />
      </React.Fragment>
      <React.Fragment>
        <TableRow
          activeRow=""
          className=""
          column={
            Array [
              Object {
                "heading": "Nama",
                "value": "name",
              },
              Object {
                "heading": "Email",
                "value": "email",
              },
              Object {
                "heading": "No. Handphone",
                "value": "phoneNumber",
              },
            ]
          }
          handleClick={[Function]}
          identifier="id"
          idx={1}
          item={
            Object {
              "email": "test@a.id",
              "name": "Test",
              "phoneNumber": "0811111111",
            }
          }
          meta={Object {}}
        />
      </React.Fragment>
    </tbody>
  </table>
</React.Fragment>
`;

exports[`src/components/elements/DataTable render with children 1`] = `
<React.Fragment>
  <table
    className=""
  >
    <thead>
      <tr>
        <th />
        <TableHeader
          item={
            Object {
              "heading": "Nama",
              "value": "name",
            }
          }
        />
        <TableHeader
          item={
            Object {
              "heading": "Email",
              "value": "email",
            }
          }
        />
        <TableHeader
          item={
            Object {
              "heading": "No. Handphone",
              "value": "phoneNumber",
            }
          }
        />
      </tr>
    </thead>
    <tbody>
      <React.Fragment>
        <TableRow
          activeRow=""
          className=""
          column={
            Array [
              Object {
                "heading": "Nama",
                "value": "name",
              },
              Object {
                "heading": "Email",
                "value": "email",
              },
              Object {
                "heading": "No. Handphone",
                "value": "phoneNumber",
              },
            ]
          }
          handleClick={[Function]}
          identifier="id"
          idx={0}
          item={
            Object {
              "email": "test@a.id",
              "name": "Test",
              "phoneNumber": "0811111111",
            }
          }
          meta={Object {}}
        >
          <div />
        </TableRow>
      </React.Fragment>
      <React.Fragment>
        <TableRow
          activeRow=""
          className=""
          column={
            Array [
              Object {
                "heading": "Nama",
                "value": "name",
              },
              Object {
                "heading": "Email",
                "value": "email",
              },
              Object {
                "heading": "No. Handphone",
                "value": "phoneNumber",
              },
            ]
          }
          handleClick={[Function]}
          identifier="id"
          idx={1}
          item={
            Object {
              "email": "test@a.id",
              "name": "Test",
              "phoneNumber": "0811111111",
            }
          }
          meta={Object {}}
        >
          <div />
        </TableRow>
      </React.Fragment>
    </tbody>
  </table>
</React.Fragment>
`;
